projects/driver_controls/src/steering_wheel_io.c

```diff
@@ -0,0 +1,97 @@
+#include <stddef.h>
+#include <stdio.h>
+
+#include "driver_io.h"
+#include "gpio_expander.h"
+#include "gpio_it.h"
+#include "input_event.h"
+#include "interrupt_def.h"
+#include "steering_wheel_io.h"
+
+// Steeing wheel input identifiers
+typedef enum {
+  STEERING_WHEEL_IO_TURN_SIGNAL = 0,
+  STEERING_WHEEL_IO_CRUISE_CONTROL,
+  STEERING_WHEEL_IO_CRUISE_CONTROL_INC,
+  STEERING_WHEEL_IO_CRUISE_CONTROL_DEC,
+  STEERING_WHEEL_IO_PUSH_TO_TALK,
+  STEERING_WHEEL_IO_BMS_FAULT,
+  STEERING_WHEEL_IO_HORN
+} SteeringWheelIODevice;
+
+// Store the id of the device as well as the id of the event the device raises
+typedef struct SteeringWheelIOData {
+  SteeringWheelIODevice id;
+  InputEvent event;
+  InterruptEdge edge;
+} SteeringWheelIOData;
+
+static SteeringWheelIOData s_steering_input_data[NUM_GPIO_EXPANDER_PINS];
+
+static void prv_steering_wheel_callback(GPIOExpanderPin pin, GPIOState state, void *context) {
+  SteeringWheelIOData *data = (SteeringWheelIOData *)context;
+
+  gpio_expander_get_state(pin, &state);
+
+  // Since the GPIO Expander causes an interrupt on both rising and falling edge, interrupts on
+  // the incorrect edge must be ignored
+  bool rising_edge_fail = (data->edge == INTERRUPT_EDGE_RISING) && (state != GPIO_STATE_HIGH);
+  bool falling_edge_fail = (data->edge == INTERRUPT_EDGE_FALLING) && (state != GPIO_STATE_LOW);
+
+  if (rising_edge_fail || falling_edge_fail) {
+    return;
+  }
+
+  // Generate the event based on the input pin
+  uint8_t event_id = data->event;
+
+  if (data->id == STEERING_WHEEL_IO_TURN_SIGNAL && state == GPIO_STATE_LOW) {
+    event_id = INPUT_EVENT_TURN_SIGNAL_NONE;
+  }
+
+  event_raise(event_id, 0);
+  printf("Pin %d\n", pin);
+  return;
+}
+
+// Configure gpio_expander pins
+void steering_wheel_io_init(void) {
+  // Define array to store configuration settings for each pin
+  s_steering_input_data[DRIVER_IO_RIGHT_TURN_SIGNAL_PIN] =
```

Comment:
Any reason not for a `const` static array? This doesn't seem to need modification.

Reply:
Done